File: android/core/src/main/kotlin/com/theformofvoid/voidbridge/core/Protocol.kt
package com.theformofvoid.voidbridge.core

import org.json.JSONArray
import org.json.JSONObject
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException
import java.net.Socket
import java.nio.ByteBuffer
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64
import javax.crypto.AEADBadTagException
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * The VoidBridge wire protocol, version 2. Byte-compatible with the Go code in
 * internal/protocol; docs/PROTOCOL.md is the reference.
 */
object Protocol {
    const val VERSION = 2
    const val MAX_FRAME = 40 shl 20
    const val MAX_TEXT = 1 shl 20
    const val MAX_IMAGE = 25 shl 20
    const val PEER_PORT = 47829
    const val DISCOVERY_PORT = 47830
    const val SERVER_PORT = 47831
    const val NONCE_SIZE = 16
    private const val PBKDF_ITERATIONS = 200_000

    const val DIR_CLIENT = 0x76620001
    const val DIR_SERVER = 0x76620002

    const val CLIP_TEXT = "text"
    const val CLIP_IMAGE = "image"

    private val random = SecureRandom()

    fun normalizeCode(code: String): String = buildString {
        for (c in code.uppercase()) {
            when (c) {
                '0' -> append('O')
                '1' -> append('I')
                '8' -> append('B')
                in 'A'..'Z', in '2'..'7' -> append(c)
            }
        }
    }

    fun validCode(code: String) = normalizeCode(code).length == 16

    fun formatCode(code: String): String {
        val n = normalizeCode(code)
        return if (n.length != 16) code else n.chunked(4).joinToString("-")
    }

    fun newGroupCode(): String {
        val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
        return formatCode(String(CharArray(16) { alphabet[random.nextInt(32)] }))
    }

    fun normalizeUsername(u: String) = u.trim().lowercase()

    private fun pbkdf(secret: String, salt: String): ByteArray {
        val spec = PBEKeySpec(secret.toCharArray(), salt.toByteArray(), PBKDF_ITERATIONS, 256)
        return SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256").generateSecret(spec).encoded
    }

    /** Slow on purpose; call off the main thread. */
    fun masterFromCode(code: String) = pbkdf(normalizeCode(code), "voidbridge-pairing-v1")

    /** Slow on purpose; call off the main thread. */
    fun masterFromAccount(username: String, password: String) =
        pbkdf(password, "voidbridge-account-v1:" + normalizeUsername(username))

    fun hmac(key: ByteArray, vararg parts: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        parts.forEach { mac.update(it) }
        return mac.doFinal()
    }

    /** HKDF-SHA256 (RFC 5869) with an empty salt, for one 32-byte output. */
    fun hkdf(secret: ByteArray, info: String): ByteArray {
        val prk = hmac(ByteArray(32), secret)
        return hmac(prk, info.toByteArray(), byteArrayOf(1))
    }

    fun sessionKey(linkKey: ByteArray, clientNonce: ByteArray, serverNonce: ByteArray) =
        hmac(linkKey, "voidbridge-session-v2".toByteArray(), clientNonce, serverNonce)

    fun newNonce(): ByteArray = ByteArray(NONCE_SIZE).also { random.nextBytes(it) }
    fun newId(): String = ByteArray(8).also { random.nextBytes(it) }.toHex()

    fun hash(kind: String, data: ByteArray): String {
        val md = MessageDigest.getInstance("SHA-256")
        md.update(kind.toByteArray())
        md.update(0)
        md.update(data)
        return md.digest().toHex()
    }

    // ---- end-to-end clip encryption ----

    private fun clipAAD(m: Message) =
        "voidbridge-clip-v2|${m.id}|${m.origin}|${m.time}|${m.clipType}|${m.mime}".toByteArray()

    fun sealClip(contentKey: ByteArray, m: Message, plaintext: ByteArray): Message {
        val nonce = ByteArray(12).also { random.nextBytes(it) }
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        c.init(Cipher.ENCRYPT_MODE, SecretKeySpec(contentKey, "AES"), GCMParameterSpec(128, nonce))
        c.updateAAD(clipAAD(m))
        return m.copy(body = nonce + c.doFinal(plaintext))
    }

    fun openClip(contentKey: ByteArray, m: Message): ByteArray? {
        val body = m.body ?: return null
        if (body.size < 12 + 16) return null
        return try {
            val c = Cipher.getInstance("AES/GCM/NoPadding")
            c.init(Cipher.DECRYPT_MODE, SecretKeySpec(contentKey, "AES"), GCMParameterSpec(128, body, 0, 12))
            c.updateAAD(clipAAD(m))
            c.doFinal(body, 12, body.size - 12)
        } catch (e: Exception) {
            null
        }
    }
}

fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }
fun String.hexToBytes(): ByteArray = ByteArray(length / 2) { substring(it * 2, it * 2 + 2).toInt(16).toByte() }

/** Everything derived from a group secret. */
class Keys(val master: ByteArray) {
    val link = Protocol.hkdf(master, "voidbridge-link-v2")
    val content = Protocol.hkdf(master, "voidbridge-content-v2")
    val auth = Protocol.hkdf(master, "voidbridge-auth-v2")
    val fingerprint: String = Protocol.hmac(link, "voidbridge-beacon-v2".toByteArray()).copyOf(8).toHex()
}

data class Identity(val id: String, val name: String, val kind: String = "android")

data class PeerInfo(val id: String, val name: String = "", val kind: String = "", val addrs: List<String> = emptyList()) {
    fun toJson(): JSONObject = JSONObject().put("id", id).apply {
        if (name.isNotEmpty()) put("name", name)
        if (kind.isNotEmpty()) put("kind", kind)
        if (addrs.isNotEmpty()) put("addrs", JSONArray(addrs))
    }

    companion object {
        fun fromJson(o: JSONObject) = PeerInfo(
            o.optString("id"), o.optString("name"), o.optString("kind"),
            o.optJSONArray("addrs")?.let { a -> List(a.length()) { a.optString(it) } } ?: emptyList(),
        )
    }
}

/** A frame header plus optional binary body. Field names match the Go JSON tags. */
data class Message(
    val type: String,
    val version: Int = 0,
    val id: String = "",
    val name: String = "",
    val kind: String = "",
    val nonce: ByteArray? = null,
    val addrs: List<String> = emptyList(),
    val origin: String = "",
    val originName: String = "",
    val time: Long = 0,
    val clipType: String = "",
    val mime: String = "",
    val peers: List<PeerInfo> = emptyList(),
    val body: ByteArray? = null,
) {
    /** Newer by copy time, then id, exactly as the Go side orders clips. */
    fun newerThan(cur: Message?): Boolean =
        cur == null || time > cur.time || (time == cur.time && id > cur.id)

    fun encode(): ByteArray {
        val o = JSONObject().put("t", type)
        if (version != 0) o.put("v", version)
        if (id.isNotEmpty()) o.put("id", id)
        if (name.isNotEmpty()) o.put("name", name)
        if (kind.isNotEmpty()) o.put("kind", kind)
        if (nonce != null) o.put("nonce", Base64.getEncoder().encodeToString(nonce))
        if (addrs.isNotEmpty()) o.put("addrs", JSONArray(addrs))
        if (origin.isNotEmpty()) o.put("origin", origin)
        if (originName.isNotEmpty()) o.put("origin_name", originName)
        if (time != 0L) o.put("time", time)
        if (clipType.isNotEmpty()) o.put("ctype", clipType)
        if (mime.isNotEmpty()) o.put("mime", mime)
        if (peers.isNotEmpty()) o.put("peers", JSONArray(peers.map { it.toJson() }))
        val h = o.toString().toByteArray(Charsets.UTF_8)
        val b = body ?: ByteArray(0)
        return ByteBuffer.allocate(4 + h.size + b.size).putInt(h.size).put(h).put(b).array()
    }

    companion object {
        const val HELLO = "hello"
        const val READY = "ready"
        const val PING = "ping"
        const val PONG = "pong"
        const val CLIP = "clip"
        const val PEERS = "peers"
        const val DEVICES = "devices"

        fun decode(b: ByteArray): Message {
            if (b.size < 4) throw IOException("short frame")
            val n = ByteBuffer.wrap(b).int
            if (n < 0 || n > b.size - 4) throw IOException("bad header length")
            val o = try {
                JSONObject(String(b, 4, n, Charsets.UTF_8))
            } catch (e: Exception) {
                throw IOException("bad header", e)
            }
            val t = o.optString("t")
            if (t.isEmpty()) throw IOException("message without type")
            fun strings(key: String) = o.optJSONArray(key)?.let { a -> List(a.length()) { a.optString(it) } } ?: emptyList()
            return Message(
                type = t,
                version = o.optInt("v"),
                id = o.optString("id"),
                name = o.optString("name"),
                kind = o.optString("kind"),
                nonce = o.optString("nonce").takeIf { it.isNotEmpty() }?.let { Base64.getDecoder().decode(it) },
                addrs = strings("addrs"),
                origin = o.optString("origin"),
                originName = o.optString("origin_name"),
                time = o.optLong("time"),
                clipType = o.optString("ctype"),
                mime = o.optString("mime"),
                peers = o.optJSONArray("peers")?.let { a -> List(a.length()) { PeerInfo.fromJson(a.getJSONObject(it)) } } ?: emptyList(),
                body = if (b.size > 4 + n) b.copyOfRange(4 + n, b.size) else null,
            )
        }
    }
}

class AuthException : IOException("devices are not in the same group")

/** AES-256-GCM with direction-tagged counter nonces (direct links only). */
class FrameCipher(sessionKey: ByteArray, isServer: Boolean) {
    private val key = SecretKeySpec(sessionKey, "AES")
    private val sendDir = if (isServer) Protocol.DIR_SERVER else Protocol.DIR_CLIENT
    private val recvDir = if (isServer) Protocol.DIR_CLIENT else Protocol.DIR_SERVER
    private var sendCount = 0L
    private var recvCount = 0L

    private fun nonce(dir: Int, n: Long) = ByteBuffer.allocate(12).putInt(dir).putLong(n).array()

    fun seal(plain: ByteArray): ByteArray {
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        c.init(Cipher.ENCRYPT_MODE, key, GCMParameterSpec(128, nonce(sendDir, sendCount++)))
        return c.doFinal(plain)
    }

    fun open(sealed: ByteArray): ByteArray {
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        c.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(128, nonce(recvDir, recvCount)))
        val p = try {
            c.doFinal(sealed)
        } catch (e: AEADBadTagException) {
            throw AuthException()
        }
        recvCount++
        return p
    }
}

/** An authenticated, encrypted direct link to another device. */
class Connection private constructor(
    private val socket: Socket,
    private val input: DataInputStream,
    private val output: DataOutputStream,
    private val cipher: FrameCipher,
    val peer: Identity,
    val peerAddrs: List<String>,
    val dialer: Boolean,
) {
    private val sendLock = Any()
    private val recvLock = Any()

    val remoteHost: String get() = socket.inetAddress?.hostAddress ?: ""
    val remoteAddr: String get() = "$remoteHost:${socket.port}"

    fun send(m: Message) {
        val plain = m.encode()
        synchronized(sendLock) { writeFrame(output, cipher.seal(plain)) }
    }

    /** Blocks until a message arrives; throws SocketTimeoutException after the socket's soTimeout of silence. */
    fun recv(): Message = synchronized(recvLock) { Message.decode(cipher.open(readFrame(input))) }

    fun setIdleTimeout(ms: Int) {
        socket.soTimeout = ms
    }

    fun close() = try {
        socket.close()
    } catch (_: IOException) {
    }

    companion object {
        fun writeFrame(out: DataOutputStream, payload: ByteArray) {
            if (payload.size > Protocol.MAX_FRAME + 64) throw IOException("frame too large")
            out.writeInt(payload.size)
            out.write(payload)
            out.flush()
        }

        fun readFrame(input: DataInputStream, max: Int = Protocol.MAX_FRAME + 64): ByteArray {
            val n = input.readInt()
            if (n < 0 || n > max) throw IOException("frame too large")
            return ByteArray(n).also { input.readFully(it) }
        }

        fun handshake(socket: Socket, linkKey: ByteArray, me: Identity, myAddrs: List<String>, dialer: Boolean, timeoutMs: Int): Connection {
            socket.soTimeout = timeoutMs
            val input = DataInputStream(socket.getInputStream().buffered(64 * 1024))
            val output = DataOutputStream(socket.getOutputStream().buffered(64 * 1024))
            val myNonce = Protocol.newNonce()
            writeFrame(output, Message(Message.HELLO, version = Protocol.VERSION, id = me.id, name = me.name, kind = me.kind, nonce = myNonce, addrs = myAddrs).encode())
            val peer = Message.decode(readFrame(input, 64 * 1024))
            if (peer.type != Message.HELLO) throw IOException("expected hello, got ${peer.type}")
            if (peer.version != Protocol.VERSION) throw IOException("the other device speaks protocol ${peer.version}; update VoidBridge on both")
            val peerNonce = peer.nonce ?: throw IOException("hello without nonce")
            if (peerNonce.size != Protocol.NONCE_SIZE || peer.id.isEmpty()) throw IOException("bad hello")
            if (peer.id == me.id) throw IOException("connected to myself")
            val (cn, sn) = if (dialer) myNonce to peerNonce else peerNonce to myNonce
            val conn = Connection(
                socket, input, output, FrameCipher(Protocol.sessionKey(linkKey, cn, sn), !dialer),
                Identity(peer.id, peer.name, peer.kind), peer.addrs, dialer,
            )
            conn.send(Message(Message.READY))
            val ready = conn.recv()
            if (ready.type != Message.READY) throw IOException("expected ready, got ${ready.type}")
            return conn
        }
    }
}

File: internal/protocol/protocol.go
// Package protocol implements the VoidBridge wire format and cryptography.
// docs/PROTOCOL.md is the reference; android/core must stay byte-compatible.
package protocol

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/hkdf"
	"crypto/hmac"
	"crypto/pbkdf2"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base32"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strconv"
	"strings"
)

const (
	Version = 2

	// MaxFrame bounds one frame (header + body, before link encryption).
	MaxFrame = 40 << 20
	MaxText  = 1 << 20
	MaxImage = 25 << 20

	PeerPort      = 47829
	DiscoveryPort = 47830
	ServerPort    = 47831

	pbkdfIterations = 200_000
	nonceSize       = 16
)

// Message types.
const (
	TypeHello   = "hello"
	TypeReady   = "ready"
	TypePing    = "ping"
	TypePong    = "pong"
	TypeClip    = "clip"
	TypePeers   = "peers"
	TypeDevices = "devices"
)

// Clip content types.
const (
	ClipText  = "text"
	ClipImage = "image"
)

// PeerInfo is a device and the addresses ("host:port") it may be reached at.
type PeerInfo struct {
	ID    string   `json:"id"`
	Name  string   `json:"name,omitempty"`
	Kind  string   `json:"kind,omitempty"`
	Addrs []string `json:"addrs,omitempty"`
}

// Message is a frame header. Body carries binary payload (encrypted clip
// content) and is not part of the JSON.
type Message struct {
	Type string `json:"t"`

	// hello: device identity. clip: ID is the clip id.
	Version int      `json:"v,omitempty"`
	ID      string   `json:"id,omitempty"`
	Name    string   `json:"name,omitempty"`
	Kind    string   `json:"kind,omitempty"` // device kind: "windows", "android"
	Nonce   []byte   `json:"nonce,omitempty"`
	Addrs   []string `json:"addrs,omitempty"`

	// clip
	Origin     string `json:"origin,omitempty"`      // device id that copied it
	OriginName string `json:"origin_name,omitempty"` // for display
	Time       int64  `json:"time,omitempty"`        // unix ms when copied
	ClipType   string `json:"ctype,omitempty"`
	Mime       string `json:"mime,omitempty"`

	// peers / devices
	Peers []PeerInfo `json:"peers,omitempty"`

	Body []byte `json:"-"`
}

// Newer reports whether clip m should replace clip cur (nil cur loses).
// Ordering is by copy time, then by id so every device agrees on ties.
func (m *Message) Newer(cur *Message) bool {
	if cur == nil {
		return true
	}
	if m.Time != cur.Time {
		return m.Time > cur.Time
	}
	return m.ID > cur.ID
}

var (
	ErrFrameTooLarge = errors.New("protocol: frame too large")
	ErrAuth          = errors.New("protocol: authentication failed (devices are not in the same group)")
	ErrDecrypt       = errors.New("protocol: cannot decrypt clip (different group key)")
)

// ---- Keys ----

// Keys are everything derived from a group secret.
type Keys struct {
	Master  []byte
	Link    []byte // authenticates and encrypts direct device-to-device links
	Content []byte // end-to-end encrypts clip contents
	Auth    []byte // proves account ownership to a server
}

// KeysFromMaster derives the subkeys.
func KeysFromMaster(master []byte) Keys {
	sub := func(label string) []byte {
		k, err := hkdf.Key(sha256.New, master, nil, label, 32)
		if err != nil {
			panic(err)
		}
		return k
	}
	return Keys{
		Master:  master,
		Link:    sub("voidbridge-link-v2"),
		Content: sub("voidbridge-content-v2"),
		Auth:    sub("voidbridge-auth-v2"),
	}
}

func pbkdf(secret, salt string) []byte {
	k, err := pbkdf2.Key(sha256.New, secret, []byte(salt), pbkdfIterations, 32)
	if err != nil {
		panic(err)
	}
	return k
}

// MasterFromCode derives the master key of a code-based (serverless) group.
func MasterFromCode(code string) []byte {
	return pbkdf(NormalizeCode(code), "voidbridge-pairing-v1")
}

// MasterFromAccount derives the master key of a server account. The password
// never leaves the device; the server only sees Keys.Auth.
func MasterFromAccount(username, password string) []byte {
	return pbkdf(password, "voidbridge-account-v1:"+NormalizeUsername(username))
}

// NormalizeUsername lowercases and trims a username.
func NormalizeUsername(u string) string { return strings.ToLower(strings.TrimSpace(u)) }

// Fingerprint identifies a group in discovery beacons without revealing keys.
func (k Keys) Fingerprint() string {
	m := hmac.New(sha256.New, k.Link)
	m.Write([]byte("voidbridge-beacon-v2"))
	return hex.EncodeToString(m.Sum(nil)[:8])
}

// NewGroupCode returns a fresh 80-bit code formatted XXXX-XXXX-XXXX-XXXX.
func NewGroupCode() string {
	b := make([]byte, 10)
	rand.Read(b)
	s := base32.StdEncoding.EncodeToString(b)
	return s[0:4] + "-" + s[4:8] + "-" + s[8:12] + "-" + s[12:16]
}

// NormalizeCode uppercases, strips separators and fixes 0/1/8 typos.
func NormalizeCode(code string) string {
	var sb strings.Builder
	for _, r := range strings.ToUpper(code) {
		switch {
		case r == '0':
			sb.WriteRune('O')
		case r == '1':
			sb.WriteRune('I')
		case r == '8':
			sb.WriteRune('B')
		case r >= 'A' && r <= 'Z', r >= '2' && r <= '7':
			sb.WriteRune(r)
		}
	}
	return sb.String()
}

// ValidCode reports whether code is a complete group code.
func ValidCode(code string) bool { return len(NormalizeCode(code)) == 16 }

// FormatCode pretty-prints a code as XXXX-XXXX-XXXX-XXXX.
func FormatCode(code string) string {
	n := NormalizeCode(code)
	if len(n) != 16 {
		return code
	}
	return n[0:4] + "-" + n[4:8] + "-" + n[8:12] + "-" + n[12:16]
}

// NewID returns a random 16-hex-digit id.
func NewID() string {
	b := make([]byte, 8)
	rand.Read(b)
	return hex.EncodeToString(b)
}

// NewNonce returns a random handshake nonce.
func NewNonce() []byte {
	b := make([]byte, nonceSize)
	rand.Read(b)
	return b
}

// ---- Clip content encryption (end to end) ----

func clipAAD(m *Message) []byte {
	return []byte("voidbridge-clip-v2|" + m.ID + "|" + m.Origin + "|" + strconv.FormatInt(m.Time, 10) + "|" + m.ClipType + "|" + m.Mime)
}

// SealClip encrypts plaintext into m.Body with the content key. The clip's
// header fields are bound as associated data so they can't be altered.
func SealClip(contentKey []byte, m *Message, plaintext []byte) {
	aead := gcm(contentKey)
	nonce := make([]byte, aead.NonceSize())
	rand.Read(nonce)
	m.Body = aead.Seal(nonce, nonce, plaintext, clipAAD(m))
}

// OpenClip decrypts m.Body.
func OpenClip(contentKey []byte, m *Message) ([]byte, error) {
	aead := gcm(contentKey)
	if len(m.Body) < aead.NonceSize() {
		return nil, ErrDecrypt
	}
	p, err := aead.Open(nil, m.Body[:aead.NonceSize()], m.Body[aead.NonceSize():], clipAAD(m))
	if err != nil {
		return nil, ErrDecrypt
	}
	return p, nil
}

func gcm(key []byte) cipher.AEAD {
	block, err := aes.NewCipher(key)
	if err != nil {
		panic(err)
	}
	aead, err := cipher.NewGCM(block)
	if err != nil {
		panic(err)
	}
	return aead
}

// ---- Frame encoding ----

// Encode serialises a message: uint32 header length, JSON header, body.
func Encode(m *Message) []byte {
	h, err := json.Marshal(m)
	if err != nil {
		panic(err)
	}
	out := make([]byte, 4+len(h)+len(m.Body))
	binary.BigEndian.PutUint32(out, uint32(len(h)))
	copy(out[4:], h)
	copy(out[4+len(h):], m.Body)
	return out
}

// Decode parses a frame produced by Encode.
func Decode(b []byte) (*Message, error) {
	if len(b) < 4 {
		return nil, errors.New("protocol: short frame")
	}
	n := binary.BigEndian.Uint32(b)
	if uint64(n) > uint64(len(b)-4) {
		return nil, errors.New("protocol: bad header length")
	}
	var m Message
	dec := json.NewDecoder(bytes.NewReader(b[4 : 4+n]))
	if err := dec.Decode(&m); err != nil {
		return nil, fmt.Errorf("protocol: bad header: %w", err)
	}
	if m.Type == "" {
		return nil, errors.New("protocol: message without type")
	}
	if rest := b[4+n:]; len(rest) > 0 {
		m.Body = rest
	}
	return &m, nil
}

// WriteFrame writes a length-prefixed blob.
func WriteFrame(w io.Writer, payload []byte) error {
	if len(payload) > MaxFrame+64 {
		return ErrFrameTooLarge
	}
	var hdr [4]byte
	binary.BigEndian.PutUint32(hdr[:], uint32(len(payload)))
	if _, err := w.Write(hdr[:]); err != nil {
		return err
	}
	_, err := w.Write(payload)
	return err
}

// ReadFrameHeader reads a frame's length prefix.
func ReadFrameHeader(r io.Reader) (int, error) {
	var hdr [4]byte
	if _, err := io.ReadFull(r, hdr[:]); err != nil {
		return 0, err
	}
	n := binary.BigEndian.Uint32(hdr[:])
	if n > MaxFrame+64 {
		return 0, ErrFrameTooLarge
	}
	return int(n), nil
}

// ---- Link cipher (direct links only) ----

// Direction tags for link nonces.
const (
	DirClient uint32 = 0x76620001
	DirServer uint32 = 0x76620002
)

// SessionKey derives a per-connection key from the link key and both nonces.
func SessionKey(linkKey, clientNonce, serverNonce []byte) []byte {
	m := hmac.New(sha256.New, linkKey)
	m.Write([]byte("voidbridge-session-v2"))
	m.Write(clientNonce)
	m.Write(serverNonce)
	return m.Sum(nil)
}

// Cipher seals frames in one direction and opens them in the other, with
// counter nonces the receiver requires in exact order (no replay/reorder).
type Cipher struct {
	aead               cipher.AEAD
	sendDir, recvDir   uint32
	sendCount, recvCnt uint64
}

// NewCipher builds a Cipher; the dialer is the "client".
func NewCipher(sessionKey []byte, isServer bool) *Cipher {
	c := &Cipher{aead: gcm(sessionKey), sendDir: DirClient, recvDir: DirServer}
	if isServer {
		c.sendDir, c.recvDir = DirServer, DirClient
	}
	return c
}

func linkNonce(dir uint32, n uint64) []byte {
	b := make([]byte, 12)
	binary.BigEndian.PutUint32(b, dir)
	binary.BigEndian.PutUint64(b[4:], n)
	return b
}

// Seal encrypts one frame. Not safe for concurrent use.
func (c *Cipher) Seal(plain []byte) []byte {
	out := c.aead.Seal(nil, linkNonce(c.sendDir, c.sendCount), plain, nil)
	c.sendCount++
	return out
}

// Open decrypts the next frame. Not safe for concurrent use.
func (c *Cipher) Open(sealed []byte) ([]byte, error) {
	p, err := c.aead.Open(nil, linkNonce(c.recvDir, c.recvCnt), sealed, nil)
	if err != nil {
		return nil, ErrAuth
	}
	c.recvCnt++
	return p, nil
}

// Hash returns a hex SHA-256, used to recognise identical clipboard content.
func Hash(kind string, data []byte) string {
	h := sha256.New()
	h.Write([]byte(kind))
	h.Write([]byte{0})
	h.Write(data)
	return hex.EncodeToString(h.Sum(nil))
}
